pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go

```diff
@@ -33,6 +36,55 @@ type Waypoint struct {
 	Addresses         []netip.Addr
 }
 
+func fetchWaypoint(ctx krt.HandlerContext, Waypoints krt.Collection[Waypoint], Namespaces krt.Collection[*v1.Namespace], o metav1.ObjectMeta) *Waypoint {
+	// try fetching the waypoint defined on the object itself
+	wp := getUseWaypoint(o)
+	if wp != nil {
+		// plausible the object has a waypoint defined but that waypoint's underlying gateway is not ready, in this case we'd return nil here even if
+		// the namespace-defined waypoint is ready and would not be nil... is this OK or should we handle that? Could lead to odd behavior when
+		// o was reliant on the namespace waypoint and then get's a use-waypoint annotation added before that gateway is ready.
+		// goes from having a waypoint to having no waypoint and then eventualy gets a waypoint back
+		return krt.FetchOne[Waypoint](ctx, Waypoints, krt.FilterName(wp.Name, wp.Namespace))
+	}
+
+	// try fetching the namespace-defined waypoint
+	namespace := krt.FetchOne[*v1.Namespace](ctx, Namespaces, krt.FilterKey(o.Namespace))
```

Comment:
@howardjohn, is this type of conditional fetching a firm No in krt? we conditionally depend on the namespace only when required vs always taking the dependency.

may not be relevant in the end, I need to check the details of defining `#none` on your namespace and attempting to override that at the workload/service level.

Reply:
I think its fine as long as its based on inputs known to krt. Which it is . So if Service changes, it will recompute and build the new dependencies.

Reply:
Thank you!